### ISSUE TITLE
Development on nixos

### ISSUE THREAD

**Issue by twitchyliquid64** [linebender/xilem#66]
Not sure where this goes, but just wanted to share that I got xilem building / working on nixos. You just need to jump into a `nix-shell` with the following `shell.nix` in your tree:

```nix
let
  pkgs = import <nixpkgs> {};

  fenix = pkgs.callPackage
    (pkgs.fetchFromGitHub {
      owner = "nix-community";
      repo = "fenix";
      # commit from: 2023-03-03
      rev = "e2ea04982b892263c4d939f1cc3bf60a9c4deaa1";
      hash = "sha256-AsOim1A8KKtMWIxG+lXh5Q4P2bhOZjoUhFWJ1EuZNNk=";
    })
    { };

in pkgs.mkShell rec {

  buildInputs = with pkgs; [
    libxkbcommon xorg.libxcb libclang
    wayland wayland-protocols vulkan-loader
  ];

  nativeBuildInputs = with pkgs; [
    # Note: to use stable, just replace `default` with `stable`
    fenix.default.toolchain

    pkgs.rustPlatform.bindgenHook
    wayland-scanner
    pkg-config
  ];

  RUST_SRC_PATH = "${fenix.complete.rust-src}/lib/rustlib/src/rust/library";
  LD_LIBRARY_PATH = pkgs.lib.makeLibraryPath buildInputs;
  LIBCLANG_PATH = "${pkgs.libclang.lib}/lib";
}
```

**Comment by zoechi** [comment]
Wouldn't this be nice to have this committed to the repo so that it becomes easy to enable with with direnv?

**Comment by SpiderUnderUrBed** [comment]
At the very least this should be in the documentation

**Comment by DJMcNab** [comment]
Our platform support documentation could definitely use some work. I agree with @SpiderUnderUrBed that putting this at the repository root is not the correct approach; it suggests that we have checked the contents, which isn't the case.

My preference would probably be to just say that we're likely be a subset of the dependencies in https://github.com/bevyengine/bevy/blob/main/docs/linux_dependencies.md, and defer doing these docs properly for now.

**Comment by PoignardAzur** [comment]
Feels like we might want to add the nix file somewhere in the `masonry_winit` folder and point to it in the `xilem` crate as well?

**Comment by PoignardAzur** [comment]
Alright. I don't feel confident enough with Nix to write one, but if someone wants to write a PR that adds such a top-level file, I'd be happy to merge it.

**Comment by PoignardAzur** [comment]
Reading that discussion:

> I think when adding it to `docs/flake.nix` (probably add `docs/flake.lock` to `.gitignore` then) and pointing to it within something that mentions linux dependencies (maybe like in bevy?), is a good compromise (i.e. doesn't make it too inconvenient for nix users, as they can just do nix develop in the docs folder, while not "committing" to the effort of having a well maintained Nix flake).

Maybe we go for that solution, and revisit it later on? If we find that the flake doesn't really bitrot, I think there's a stronger argument to keep it at the top level.

**Comment by DJMcNab** [comment]
I don't mind adding such a file somewhere in our repository, but I really don't want it to be in the repository root.

The main reason for that is that it can easily become stale, because only a very few contributors care about it. And it being out-of-date would lead to a worse user experience. For example, what happens when we update our MSRV? If our Nix setup picks an out-of-date Rust version automatically, then that user's first impression of our project is one of carelessless.

I don't want this to have a CI check, because it means that people's work can be blocked on a file which only a subset of users can meaningfully debug - I don't think asking people to learn how to use Nix is a productive use of limited contributor time.
Or if it's not blocking, GitHub will still report the CI as "failing" on PRs, even when they aren't - which still makes for a poor contributor experience - I think it's unavoidable that contributors would try and fix it. And it would also give the impression that even though we have CI, we don't keep it clean, even though its for a peripheral thing.

If we can tell these contributors that they can run `cp docs/flake.nix flake.nix` to get it working automatically, then that means we aren't punishing the post-setup UX. (That is, I think it would be correct to add `/flake.nix` to `.gitignore` if we did that).
But I want a process that requires a manual step, so we can inform them that this file is best-effort only, and that it's not maintained by us.

**Comment by Philipp-M** [comment]
> If our Nix setup picks an out-of-date Rust version automatically, then that user's first impression of our project is one of carelessless.

We could automatically use the version that's MSRV, in there. Though that still requires that the `flake.lock` is somewhat up-to-date. I think I could maintain it (though no promises).

> I don't want this to have a CI check

Yeah I don't think that this was currently the goal (of this issue at least), although it really enables reproducible builds. This can always later be done if it really makes sense (but I agree, that Nix is still somewhat niche and fixing nix flakes can sometimes be non-trivial unfortunately).

> If we can tell these contributors that they can run cp docs/flake.nix flake.nix to get it working automatically

I don't think that's necessary. It should be possible to just `nix develop` within `docs/`.
And it should also be possible to use it as flake input as well (just not directly pointing to the root of the repo), whether that is desired is a different issue (I don't think it's meaningful currently, as it's a library, for applications that would be different).

Ok, right than I'll add it to `docs/` with just `devShells` for now, we can always later extend it, and that should be sufficient to close this issue.
